JS: recognize sanitizing slashes in URL redirection queries

asger-semmle opened the pull request:
The sanitizer for our URL redirection queries is now more precise, so we no longer get false positives from things like:
```javascript
res.redirect("/users/" + req.params.id);
```

The tricky thing about slashes is that they are used in multiple parts of the URL:

1. separating the (optional) scheme from the hostname
2. separating the hostname from the path
3. occurring inside the path, query, and/or fragment

The attacker must not be able to control the hostname, so we need to make sure a slash cannot be interpreted as (1). A redirection such as
```javascript
res.redirect("/" + req.query.redirect)
```
is unsafe because it can lead to `//evil.com`.

I've also removed the `Sink.maybeNonLocal()` predicate from ServerSideUrlRedirect as it appears to implement the same sanitizer as the sanitizing prefix.

I'll run an evaluation, but just putting up for review now.

esben-semmle commented on javascript/ql/src/semmle/javascript/security/dataflow/UrlConcatenation.qll:
```diff
@@ -7,11 +7,19 @@
-private predicate hasSanitizingSubstring(DataFlow::Node nd) {
-  nd.asExpr().getStringValue().regexpMatch(".*[?#].*")
+predicate hasSanitizingSubstring(DataFlow::Node nd) {
+  nd.asExpr().getStringValue().regexpMatch(".*([?#]|[^?#:/\\\\][/\\\\]).*") 
```
The `?#` in the negated class is not needed, is it?

asger-semmle commented:
[Performance](https://git.semmle.com/gist/asger/fd47dc60d4a9a6368f38ba561777d61e) goes both ways. I suspect the difference is that there are more sinks now (unreachable due to the sanitizer), so the exploratory flow steps reach a larger graph.

asger-semmle commented:
I had also missed that the SSRF query depends on `sanitizingPrefixEdge`, and for SSRF the path is also considered vulnerable. So I've reverted `sanitizingPrefixEdge` to its original behaviour and introduced `hostnameSanitizingPrefixEdge` instead.

xiemaisi commented:
Conflicts again.
